### ISSUE TITLE
Virtuoso RDFZZ Error DB.DBA.SPARQL_REXEC unsupported Content-Type 'application/json'

### ISSUE THREAD

**Issue by holycrab13** [openlink/virtuoso-opensource#1005]
`Virtuoso RDFZZ Error DB.DBA.SPARQL_REXEC('[MY ENDPOINT], ...) returned unsupported Content-Type 'application/json'`

when executing a federated query that does local aggregation (`GROUP CONCAT`) over the federated result.
This is the query:

```
PREFIX dataid: <http://dataid.dbpedia.org/ns/core#>
PREFIX dcv: <http://dataid.dbpedia.org/ns/cv#>
PREFIX dct:    <http://purl.org/dc/terms/>
PREFIX dcat:   <http://www.w3.org/ns/dcat#>
PREFIX rdf:    <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs:   <http://www.w3.org/2000/01/rdf-schema#>
SELECT DISTINCT ?versionUri ?dataset ?distribution ?label ?comment (GROUP_CONCAT(DISTINCT ?file; SEPARATOR=", ") AS ?files) ?license ?size ?version ?format (GROUP_CONCAT(DISTINCT ?var; SEPARATOR=', ') AS ?variant) ?preview WHERE
{
	{
		SERVICE <https://dev.databus.dbpedia.org/sparql>
		{
			GRAPH ?g
			{
				{
					?dataset dataid:group <https://dev.databus.dbpedia.org/janfo/dbpedia> .
					{
						?dataset dataid:artifact <https://dev.databus.dbpedia.org/janfo/dbpedia/geo-coordinates-mappingbased> .
						{
							?distribution <http://dataid.dbpedia.org/ns/cv#lang> ?c0 .
							VALUES ?c0 {
								'en'
								'fr'
							}
						}
					}
				}
				?distribution dataid:file ?file .
				?distribution dataid:formatExtension ?format .
				OPTIONAL { ?distribution ?p  ?var. ?p rdfs:subPropertyOf dataid:contentVariant . }
				OPTIONAL { ?dataset dct:license ?license . }
				OPTIONAL { ?distribution dcat:byteSize ?size . }
				OPTIONAL { ?distribution dataid:preview ?preview . }
				?dataset dcat:distribution ?distribution .
				?dataset dcat:distribution ?distribution .
				?dataset dataid:version ?versionUri .
				?dataset dct:hasVersion ?version .
				?dataset dct:title ?label .
				?dataset dct:abstract ?comment.
			}
		}
	}
}
GROUP BY ?versionUri ?dataset ?distribution ?label ?comment ?license ?size ?version ?format ?preview
```

Running the same query with only one result works (e.g., removing the `'fr'` from the `VALUES` clause)

**Comment by holycrab13** [comment]
Is most likely caused by a proxy not setting the request header correctly when redirecting the SPARQL request, issue can be closed

**Comment by kidehen** [comment]
[This SPARQL-FED variant](http://linkeddata.uriburner.com/sparql/?default-graph-uri=&query=PREFIX+dataid%3A+%3Chttp%3A%2F%2Fdataid.dbpedia.org%2Fns%2Fcore%23%3E%0D%0APREFIX+dcv%3A+%3Chttp%3A%2F%2Fdataid.dbpedia.org%2Fns%2Fcv%23%3E%0D%0APREFIX+dct%3A++++%3Chttp%3A%2F%2Fpurl.org%2Fdc%2Fterms%2F%3E%0D%0APREFIX+dcat%3A+++%3Chttp%3A%2F%2Fwww.w3.org%2Fns%2Fdcat%23%3E%0D%0APREFIX+rdf%3A++++%3Chttp%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%3E%0D%0APREFIX+rdfs%3A+++%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0D%0ASELECT+DISTINCT+%3FversionUri+%3Fdataset+%3Fdistribution+%3Flabel+%3Fcomment+%28GROUP_CONCAT%28DISTINCT+%3Ffile%3B+SEPARATOR%3D%22%2C+%22%29+AS+%3Ffiles%29+%3Flicense+%3Fsize+%3Fversion+%3Fformat+%28GROUP_CONCAT%28DISTINCT+%3Fvar%3B+SEPARATOR%3D%27%2C+%27%29+AS+%3Fvariant%29+%3Fpreview+WHERE%0D%0A%7B%0D%0A%09%7B%0D%0A%09%09SERVICE+%3Chttps%3A%2F%2Fdev.databus.dbpedia.org%2Fsparql%3E%0D%0A%09%09%7B%0D%0A+++++++++++++++++++SELECT+*+%0D%0A+++++++++++++++++++WHERE+%7B%0D%0A%09%09%09GRAPH+%3Fg%0D%0A%09%09%09%7B%0D%0A%09%09%09%09%7B%0D%0A%09%09%09%09%09%3Fdataset+dataid%3Agroup+%3Chttps%3A%2F%2Fdev.databus.dbpedia.org%2Fjanfo%2Fdbpedia%3E+.%0D%0A%09%09%09%09%09%7B%0D%0A%09%09%09%09%09%09%3Fdataset+dataid%3Aartifact+%3Chttps%3A%2F%2Fdev.databus.dbpedia.org%2Fjanfo%2Fdbpedia%2Fgeo-coordinates-mappingbased%3E+.%0D%0A%09%09%09%09%09%09%7B%0D%0A%09%09%09%09%09%09%09%3Fdistribution+%3Chttp%3A%2F%2Fdataid.dbpedia.org%2Fns%2Fcv%23lang%3E+%3Fc0+.%0D%0A%09%09%09%09%09%09%09VALUES+%3Fc0+%7B%27en%27%27fr%27%7D%0D%0A%09%09%09%09%09%09%7D%0D%0A%09%09%09%09%09%7D%0D%0A%09%09%09%09%7D%0D%0A%09%09%09%09%3Fdistribution+dataid%3Afile+%3Ffile+.%0D%0A%09%09%09%09%3Fdistribution+dataid%3AformatExtension+%3Fformat+.%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdistribution+%3Fp++%3Fvar.+%3Fp+rdfs%3AsubPropertyOf+dataid%3AcontentVariant+.+%7D%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdataset+dct%3Alicense+%3Flicense+.+%7D%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdistribution+dcat%3AbyteSize+%3Fsize+.+%7D%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdistribution+dataid%3Apreview+%3Fpreview+.+%7D%0D%0A%09%09%09%09%3Fdataset+dcat%3Adistribution+%3Fdistribution+.%0D%0A%09%09%09%09%3Fdataset+dcat%3Adistribution+%3Fdistribution+.%0D%0A%09%09%09%09%3Fdataset+dataid%3Aversion+%3FversionUri+.%0D%0A%09%09%09%09%3Fdataset+dct%3AhasVersion+%3Fversion+.%0D%0A%09%09%09%09%3Fdataset+dct%3Atitle+%3Flabel+.%0D%0A%09%09%09%09%3Fdataset+dct%3Aabstract+%3Fcomment.%0D%0A%09%09%09%7D%0D%0A%09%09%7D%0D%0A+++++++++++++%7D%0D%0A%09%7D%0D%0A%7D%0D%0AGROUP+BY+%3FversionUri+%3Fdataset+%3Fdistribution+%3Flabel+%3Fcomment+%3Flicense+%3Fsize+%3Fversion+%3Fformat+%3Fpreview%0D%0A&should-sponge=&format=text%2Fhtml&timeout=30000000) works from our live URIBurner instance. 

Query Text.

```
PREFIX dataid: <http://dataid.dbpedia.org/ns/core#>
PREFIX dcv: <http://dataid.dbpedia.org/ns/cv#>
PREFIX dct:    <http://purl.org/dc/terms/>
PREFIX dcat:   <http://www.w3.org/ns/dcat#>
PREFIX rdf:    <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs:   <http://www.w3.org/2000/01/rdf-schema#>
SELECT DISTINCT ?versionUri ?dataset ?distribution ?label ?comment (GROUP_CONCAT(DISTINCT ?file; SEPARATOR=", ") AS ?files) ?license ?size ?version ?format (GROUP_CONCAT(DISTINCT ?var; SEPARATOR=', ') AS ?variant) ?preview WHERE
{
	{
		SERVICE <https://dev.databus.dbpedia.org/sparql>
		{
                   SELECT * 
                   WHERE {
			GRAPH ?g
			{
				{
					?dataset dataid:group <https://dev.databus.dbpedia.org/janfo/dbpedia> .
					{
						?dataset dataid:artifact <https://dev.databus.dbpedia.org/janfo/dbpedia/geo-coordinates-mappingbased> .
						{
							?distribution <http://dataid.dbpedia.org/ns/cv#lang> ?c0 .
							VALUES ?c0 {'en''fr'}
						}
					}
				}
				?distribution dataid:file ?file .
				?distribution dataid:formatExtension ?format .
				OPTIONAL { ?distribution ?p  ?var. ?p rdfs:subPropertyOf dataid:contentVariant . }
				OPTIONAL { ?dataset dct:license ?license . }
				OPTIONAL { ?distribution dcat:byteSize ?size . }
				OPTIONAL { ?distribution dataid:preview ?preview . }
				?dataset dcat:distribution ?distribution .
				?dataset dcat:distribution ?distribution .
				?dataset dataid:version ?versionUri .
				?dataset dct:hasVersion ?version .
				?dataset dct:title ?label .
				?dataset dct:abstract ?comment.
			}
		}
             }
	}
}
GROUP BY ?versionUri ?dataset ?distribution ?label ?comment ?license ?size ?version ?format ?preview
```

**Comment by TallTed** [comment]
[The original query](https://linkeddata.uriburner.com/sparql/?default-graph-uri=&qtxt=PREFIX+dataid%3A+%3Chttp%3A%2F%2Fdataid.dbpedia.org%2Fns%2Fcore%23%3E%0D%0APREFIX+dcv%3A+%3Chttp%3A%2F%2Fdataid.dbpedia.org%2Fns%2Fcv%23%3E%0D%0APREFIX+dct%3A++++%3Chttp%3A%2F%2Fpurl.org%2Fdc%2Fterms%2F%3E%0D%0APREFIX+dcat%3A+++%3Chttp%3A%2F%2Fwww.w3.org%2Fns%2Fdcat%23%3E%0D%0APREFIX+rdf%3A++++%3Chttp%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%3E%0D%0APREFIX+rdfs%3A+++%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0D%0ASELECT+DISTINCT+%3FversionUri+%3Fdataset+%3Fdistribution+%3Flabel+%3Fcomment+%28GROUP_CONCAT%28DISTINCT+%3Ffile%3B+SEPARATOR%3D%22%2C+%22%29+AS+%3Ffiles%29+%3Flicense+%3Fsize+%3Fversion+%3Fformat+%28GROUP_CONCAT%28DISTINCT+%3Fvar%3B+SEPARATOR%3D%27%2C+%27%29+AS+%3Fvariant%29+%3Fpreview+WHERE%0D%0A%7B%0D%0A%09%7B%0D%0A%09%09SERVICE+%3Chttps%3A%2F%2Fdev.databus.dbpedia.org%2Fsparql%3E%0D%0A%09%09%7B%0D%0A%09%09%09GRAPH+%3Fg%0D%0A%09%09%09%7B%0D%0A%09%09%09%09%7B%0D%0A%09%09%09%09%09%3Fdataset+dataid%3Agroup+%3Chttps%3A%2F%2Fdev.databus.dbpedia.org%2Fjanfo%2Fdbpedia%3E+.%0D%0A%09%09%09%09%09%7B%0D%0A%09%09%09%09%09%09%3Fdataset+dataid%3Aartifact+%3Chttps%3A%2F%2Fdev.databus.dbpedia.org%2Fjanfo%2Fdbpedia%2Fgeo-coordinates-mappingbased%3E+.%0D%0A%09%09%09%09%09%09%7B%0D%0A%09%09%09%09%09%09%09%3Fdistribution+%3Chttp%3A%2F%2Fdataid.dbpedia.org%2Fns%2Fcv%23lang%3E+%3Fc0+.%0D%0A%09%09%09%09%09%09%09VALUES+%3Fc0+%7B%0D%0A%09%09%09%09%09%09%09%09%27en%27%0D%0A%09%09%09%09%09%09%09%09%27fr%27%0D%0A%09%09%09%09%09%09%09%7D%0D%0A%09%09%09%09%09%09%7D%0D%0A%09%09%09%09%09%7D%0D%0A%09%09%09%09%7D%0D%0A%09%09%09%09%3Fdistribution+dataid%3Afile+%3Ffile+.%0D%0A%09%09%09%09%3Fdistribution+dataid%3AformatExtension+%3Fformat+.%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdistribution+%3Fp++%3Fvar.+%3Fp+rdfs%3AsubPropertyOf+dataid%3AcontentVariant+.+%7D%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdataset+dct%3Alicense+%3Flicense+.+%7D%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdistribution+dcat%3AbyteSize+%3Fsize+.+%7D%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdistribution+dataid%3Apreview+%3Fpreview+.+%7D%0D%0A%09%09%09%09%3Fdataset+dcat%3Adistribution+%3Fdistribution+.%0D%0A%09%09%09%09%3Fdataset+dcat%3Adistribution+%3Fdistribution+.%0D%0A%09%09%09%09%3Fdataset+dataid%3Aversion+%3FversionUri+.%0D%0A%09%09%09%09%3Fdataset+dct%3AhasVersion+%3Fversion+.%0D%0A%09%09%09%09%3Fdataset+dct%3Atitle+%3Flabel+.%0D%0A%09%09%09%09%3Fdataset+dct%3Aabstract+%3Fcomment.%0D%0A%09%09%09%7D%0D%0A%09%09%7D%0D%0A%09%7D%0D%0A%7D%0D%0AGROUP+BY+%3FversionUri+%3Fdataset+%3Fdistribution+%3Flabel+%3Fcomment+%3Flicense+%3Fsize+%3Fversion+%3Fformat+%3Fpreview%0D%0A&should-sponge=&format=text%2Fhtml&timeout=30000000) also [works on the URIBurner instance](https://linkeddata.uriburner.com/sparql/?default-graph-uri=&query=PREFIX+dataid%3A+%3Chttp%3A%2F%2Fdataid.dbpedia.org%2Fns%2Fcore%23%3E%0D%0APREFIX+dcv%3A+%3Chttp%3A%2F%2Fdataid.dbpedia.org%2Fns%2Fcv%23%3E%0D%0APREFIX+dct%3A++++%3Chttp%3A%2F%2Fpurl.org%2Fdc%2Fterms%2F%3E%0D%0APREFIX+dcat%3A+++%3Chttp%3A%2F%2Fwww.w3.org%2Fns%2Fdcat%23%3E%0D%0APREFIX+rdf%3A++++%3Chttp%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%3E%0D%0APREFIX+rdfs%3A+++%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0D%0ASELECT+DISTINCT+%3FversionUri+%3Fdataset+%3Fdistribution+%3Flabel+%3Fcomment+%28GROUP_CONCAT%28DISTINCT+%3Ffile%3B+SEPARATOR%3D%22%2C+%22%29+AS+%3Ffiles%29+%3Flicense+%3Fsize+%3Fversion+%3Fformat+%28GROUP_CONCAT%28DISTINCT+%3Fvar%3B+SEPARATOR%3D%27%2C+%27%29+AS+%3Fvariant%29+%3Fpreview+WHERE%0D%0A%7B%0D%0A%09%7B%0D%0A%09%09SERVICE+%3Chttps%3A%2F%2Fdev.databus.dbpedia.org%2Fsparql%3E%0D%0A%09%09%7B%0D%0A%09%09%09GRAPH+%3Fg%0D%0A%09%09%09%7B%0D%0A%09%09%09%09%7B%0D%0A%09%09%09%09%09%3Fdataset+dataid%3Agroup+%3Chttps%3A%2F%2Fdev.databus.dbpedia.org%2Fjanfo%2Fdbpedia%3E+.%0D%0A%09%09%09%09%09%7B%0D%0A%09%09%09%09%09%09%3Fdataset+dataid%3Aartifact+%3Chttps%3A%2F%2Fdev.databus.dbpedia.org%2Fjanfo%2Fdbpedia%2Fgeo-coordinates-mappingbased%3E+.%0D%0A%09%09%09%09%09%09%7B%0D%0A%09%09%09%09%09%09%09%3Fdistribution+%3Chttp%3A%2F%2Fdataid.dbpedia.org%2Fns%2Fcv%23lang%3E+%3Fc0+.%0D%0A%09%09%09%09%09%09%09VALUES+%3Fc0+%7B%0D%0A%09%09%09%09%09%09%09%09%27en%27%0D%0A%09%09%09%09%09%09%09%09%27fr%27%0D%0A%09%09%09%09%09%09%09%7D%0D%0A%09%09%09%09%09%09%7D%0D%0A%09%09%09%09%09%7D%0D%0A%09%09%09%09%7D%0D%0A%09%09%09%09%3Fdistribution+dataid%3Afile+%3Ffile+.%0D%0A%09%09%09%09%3Fdistribution+dataid%3AformatExtension+%3Fformat+.%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdistribution+%3Fp++%3Fvar.+%3Fp+rdfs%3AsubPropertyOf+dataid%3AcontentVariant+.+%7D%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdataset+dct%3Alicense+%3Flicense+.+%7D%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdistribution+dcat%3AbyteSize+%3Fsize+.+%7D%0D%0A%09%09%09%09OPTIONAL+%7B+%3Fdistribution+dataid%3Apreview+%3Fpreview+.+%7D%0D%0A%09%09%09%09%3Fdataset+dcat%3Adistribution+%3Fdistribution+.%0D%0A%09%09%09%09%3Fdataset+dcat%3Adistribution+%3Fdistribution+.%0D%0A%09%09%09%09%3Fdataset+dataid%3Aversion+%3FversionUri+.%0D%0A%09%09%09%09%3Fdataset+dct%3AhasVersion+%3Fversion+.%0D%0A%09%09%09%09%3Fdataset+dct%3Atitle+%3Flabel+.%0D%0A%09%09%09%09%3Fdataset+dct%3Aabstract+%3Fcomment.%0D%0A%09%09%09%7D%0D%0A%09%09%7D%0D%0A%09%7D%0D%0A%7D%0D%0AGROUP+BY+%3FversionUri+%3Fdataset+%3Fdistribution+%3Flabel+%3Fcomment+%3Flicense+%3Fsize+%3Fversion+%3Fformat+%3Fpreview%0D%0A&should-sponge=&format=text%2Fhtml&timeout=30000000) without any changes.

@holycrab13 — Sharing more details of the problematic deployment stack (such as the proxy you're suggesting is "not setting the request header correctly") may be helpful for others who hit similar errors, especially if you can share a proxy configuration, proxy version, or other change that resolves the issue.

**Comment by holycrab13** [comment]
I was using express and nodejs to implement a proxy for the SPARQL endpoint.

The GET request was just using a pipe from the `request` package and worked fine.
The POST request however had problems to forward the body and thus the query. I had to create a new request in that express route. Here I initially forgot to set the accept header of the new request to the accept header of the initial request. My configuration also specified `json: true` in the request options and thus always requested json.

This was not a problem for a long time since the federated queries where either small and probably using the GET route or my application required responses in json anyway.

Here's the fixed route for POST
```
router.post('/sparql', cors(), async function (req, res) {

    var sparqlEndpoint = `${process.env.DATABUS_DATABASE_URL}/sparql`;
    var query = req.body.query;
    var accept = req.headers['accept']

    if(accept == undefined) {
       accept = 'application/json';
    }

    var options = {
      method: 'POST',
      uri: sparqlEndpoint + '?timeout=10000',
      body: "query=" + encodeURIComponent(query),
      headers: {
        "Accept": accept,
        "Content-type": "application/x-www-form-urlencoded"
      },
    };

    request.post(options).pipe(res);
  });
```

The error message `unsupported Content-Type 'application/json'` is 100% correct, because I forced the federated query result to be returned in json, which is a problem for further query processing when doing aggregates.
